### PR TITLE
[GSoC]Fixed `_eval_subs` method for Orders

### LINKED ISSUE #19120: sin works well but cos return wrong
```
from sympy import *
cos(x).series(n=8).subs(x,cos(x).series(n=8))
```

this return type error

image

but if we use sin

```
from sympy import *
sin(x).series(n=8).subs(x,sin(x).series(n=8)).expand()
```

image

### PR DESCRIPTION
**Pull request by arnabnandikgp**
Fixes #19120
Closes #22339 
#### Brief description of what is fixed or changed

For substitution of order terms or expression containing order terms onto an order term we should be checking the asymptotic nature of the new expression and if it goes to the same point as the old then we can allow such substitution.
#### Release Notes

* series
  * Better handling of Substitution of expression containing order terms in order terms.

### CODE REVIEW

**Comment by arnabnandikgp** [comment]
ping @anutosh491 for reviews

**Comment by anutosh491** [comment]
I think this looks reasonable. Past 3 gsoc students inlcuding me for the series module have tried fixing this but the Pr was stalled due to lack of reviews. I also like this more than the approach (check #22339) I was using :)

Let's wait for @oscarbenjamin to give us the green flag !

**Comment by oscarbenjamin** [comment]
I haven't looked at `Order._eval_subs` before but now that I do it looks absolutely horrible. What on earth is all of that code doing?

A routine like `_eval_subs` should not do things like calling `limit` or `solveset` or whatever.

Is it a problem just to delete the whole method? Why is it needed?

**Comment by anutosh491** [comment]
> I haven't looked at Order._eval_subs before but now that I do it looks absolutely horrible. What on earth is all of that code doing?

Yes I get it. Wasn't my favourite thing to work on when I was looking into it couple of years ago. But answering your question, I think it is needed at some places. Mainly being this 
https://github.com/sympy/sympy/blob/72f210ce2e9c164fab5280b10df2f351a6e0dbf5/sympy/core/expr.py#L2996-L3015

So any use of the series function unless it is of the form `series(f(x), x, 0, dir = "+")` uses transformation that involves substituion (which would involve a substitution in the order term too)

But then I think not everything was clear as to how substituion works for the order class which led to issues and also use of solveset and/or limit. The solveset was used to handle the basic case of `Order(x**2).subs(x, y - 1)`

**Comment by oscarbenjamin** [comment]
> The solveset was used to handle the basic case of `Order(x**2).subs(x, y - 1)`

There has to be a better way than using `solveset`.

To be clear I don't object to this PR if it improves on the status quo.

I do object to the whole `Order._eval_subs` method though. A rethink from first principles is needed about what the method is even there for. The fact that is calling things like `limit` and `solveset` shows that something has gone badly wrong. A call to `subs` should not invoke such routines.

**Comment by oscarbenjamin** [comment]
> So any use of the series function unless it is of the form `series(f(x), x, 0, dir = "+")` uses transformation that involves substituion (which would involve a substitution in the order term too)

That should be changed. The series function should not be dependent on this sort of logic being implemented in `Order._eval_subs`. It is the job of the series function to know how to construct the `Order` correctly in the case of a substitution.

**Comment by anutosh491** [comment]
> A rethink from first principles is needed about what the method is even there for.

Absolutely correct. This is a reason I am sometimes confused if the series should even be addition of functions + an order term to make an object of the add class (rather than having its own class).

**Comment by oscarbenjamin** [comment]
> This is a reason I am sometimes confused if the series should even be addition of functions + an order term to make an object of the add class (rather than having its own class).

There is no need to be confused about this. Implementing series as an Add with an O term is a bad idea.

A Series should be a special object of its own and it should be like Poly rather than like Expr. The internal representation should be like a Poly i.e. with an actual ordered list of coefficients. The coefficients should be represented using domain elements in a canonical form, always expanded etc just like for Poly. There should be no need to waste time debating whether to call expand, cancel etc because the domain elements are always canonical.

The ring_series module shows how to do this but is too low-level (gh-26100). It needs to be wrapped in something higher-level like Poly. Also python-flint should be used when possible:
```python
In [7]: from flint import fmpq_series

In [8]: x = fmpq_series([0, 1])

In [9]: x
Out[9]: x + O(x^10)

In [10]: 1/(x + 1)
Out[10]: 1 + (-1)*x + x^2 + (-1)*x^3 + x^4 + (-1)*x^5 + x^6 + (-1)*x^7 + x^8 + (-1)*x^9 + O(x^10)

In [11]: x.exp()
Out[11]: 1 + x + 1/2*x^2 + 1/6*x^3 + 1/24*x^4 + 1/120*x^5 + 1/720*x^6 + 1/5040*x^7 + 1/40320*x^8 + 1/362880*x^9 + O(x^10)
```
It is much faster:
```python
In [1]: from flint import ctx

In [2]: ctx.cap = 1000

In [3]: from flint import fmpq_series

In [4]: x = fmpq_series([0, 1])

In [5]: %time r = x.exp()
CPU times: user 5.04 ms, sys: 79 μs, total: 5.12 ms
Wall time: 6.18 ms

In [6]: len(r)
Out[6]: 1000
```
Compare:
```python
In [12]: %time ok = exp(x).series(x, n=1000)
CPU times: user 19.2 s, sys: 23.3 ms, total: 19.3 s
Wall time: 19.3 s
```
That's 3000x slower.

It should be possible to have multivariate series as well (gh-25483).

**Comment by anutosh491** [comment]
I think you're above comment was good enough to track it as an issue :)
This should surely be addressed at some point of time.

**Comment by anutosh491** [comment]
Converting to draft as we don't have a decision to proceed yet.

**Comment by oscarbenjamin** [comment]
I said above:

> To be clear I don't object to this PR if it improves on the status quo.

I don't know whether this PR improves the state of affairs but it can be possible to merge a change that fixes one thing while still knowing that the function in question should really be rewritten in a future PR.

**Comment by anutosh491** [comment]
In that case, I like most of what I see !

File: sympy/series/tests/test_order.py [hunk 1]
```diff
@@ -421,6 +422,24 @@ def test_order_subs_limits():
...
+    assert (x**2 + O(x**2) + 1/x**2).subs(x, x + O(x**2)) == (x + O(x**2))**(-2) + O(x**2)
+    assert (x**2 + O(x**2) + 1).subs(x, x + O(x**2)) == 1 + O(x**2)
+    assert O(x, (x, oo)).subs(x, x + O(x**2, (x, oo))) == O(x**2, (x, oo))
+    assert sin(x).series(n=8).subs(x,sin(x).series(n=8)).expand() == x - x**3/3 + x**5/10 - 8*x**7/315 + O(x**8)
+    assert cos(x).series(n=8).subs(x,sin(x).series(n=8)).expand() == 1 - x**2/2 + 5*x**4/24 - 37*x**6/720 + O(x**8)
+    assert O(x).subs(x, O(1/x, (x, oo))) == O(1/x, (x, oo))
+
+@XFAIL
+def test_order_failing_due_to_solveset():
+    assert O(x**3).subs(x, exp(-x**2)) == O(exp(-3*x**2), (x, oo))
```
[the remark this answers is not in the document]
Comment (arnabnandikgp) [reply sympy/series/tests/test_order.py:440]: oh yes, this should have been `-oo`.

**Review by anutosh491 — APPROVED** [review]
LGTM

**Pull request merged**